# strip、lstrip、rstrip的使用说明

* strip() 方法用于移除字符串头尾指定的字符（默认为空格或换行符）或字符序列。
* lstrip()方法用于移除字符串左侧制定的字符（默认为空格或换行符）或字符序列。
* rstrip()方法用于移除字符串左侧制定的字符（默认为空格或换行符）或字符序列。
* 注意：该方法只能删除开头或是结尾的字符，不能删除中间部分的字符。

In [7]:
str = '000055550000'
print (str.lstrip('0'))
print (str.rstrip('0'))
print (str.strip('0'))

55550000
00005555
5555


# dir的使用说明

dir() 函数不带参数时，返回当前范围内的变量、方法和定义的类型列表；带参数时，返回参数的属性、方法列表。如果参数包含方法__dir__()，该方法将被调用。如果参数不包含__dir__()，该方法将最大限度地收集参数信息。

In [8]:
dir(str)

['__add__',
 '__class__',
 '__contains__',
 '__delattr__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__getitem__',
 '__getnewargs__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__iter__',
 '__le__',
 '__len__',
 '__lt__',
 '__mod__',
 '__mul__',
 '__ne__',
 '__new__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__rmod__',
 '__rmul__',
 '__setattr__',
 '__sizeof__',
 '__str__',
 '__subclasshook__',
 'capitalize',
 'casefold',
 'center',
 'count',
 'encode',
 'endswith',
 'expandtabs',
 'find',
 'format',
 'format_map',
 'index',
 'isalnum',
 'isalpha',
 'isdecimal',
 'isdigit',
 'isidentifier',
 'islower',
 'isnumeric',
 'isprintable',
 'isspace',
 'istitle',
 'isupper',
 'join',
 'ljust',
 'lower',
 'lstrip',
 'maketrans',
 'partition',
 'replace',
 'rfind',
 'rindex',
 'rjust',
 'rpartition',
 'rsplit',
 'rstrip',
 'split',
 'splitlines',
 'startswith',
 'strip',
 'swapcase',
 'title',
 'translate',
 'upper',
 'zfill']

# 浅谈编码

* 在Python3以后，字符串和bytes类型彻底分开了。字符串是以字符为单位进行处理的，bytes类型是以字节为单位处理的。
* 先对字符串分析。python内部是使用unicode来处理的，unicode有两种编码格式，一种是UCS-2，有65536个码位，另一种是UCS-4，有2147483648g个码位。对于这两种码位，python都是支持的，这个是在编码时通过--encode-unicode=ucs2 或 --encode-unicode=ucs4来指定的。可通过下面方式确定自己默认安装的python的编码：

In [26]:
import sys
print(sys.maxunicode)

1114111


* 如果输出的值为65535，那么就是UCS-2；如果是1114111，那么就是UCS-4。
* 对bytes分析。bytes数据类型在所有的操作和使用甚至内置方法上和字符串数据类型基本一样，也是不可变序列对象。bytes对象只负责以二进制字节序列的形式记录所需记录的对象，至于对象是什么则由相应地编码格式解码决定。Python3中，bytes通常用于网络数据传输、二进制图片和文件的保存等等。可以通过调用bytes()生成bytes，如下：

In [52]:
a = b'hello'  #直接指定你好为bytes类型
a2 = b''  #建立空bytes类型
b = bytes()    #....
print(a,a2,b)
print(type(a),type(b))

b'hello' b'' b''
<class 'bytes'> <class 'bytes'>


* 注意：创立bytes数据类型时，字符串必须为ASCII码所含有的，不能用汉语。但可以通过编码，将汉语字符串编码成bytes数据类型

In [62]:
try:
    a = bytes('你好')
except SyntaxError as e :
    print(e)
except TypeError as e:
    print(e)
else:
    print(type(a))
##方法一
import codecs
str = '你好'
look = codecs.lookup('utf-8')
byter = look.encode(str)
print(byter)
print(type(byter[0]))
##方法二
byter2 = bytes('你好', encoding = 'utf-8')
print(type(byter2))
##方法三
print(type(str.encode('utf-8')))  #默认为utf-8

string argument without an encoding
(b'\xe4\xbd\xa0\xe5\xa5\xbd', 2)
<class 'bytes'>
<class 'bytes'>
<class 'bytes'>


* 字符串类型与bytes类型的转换

In [61]:
##字符串转换为bytes类型，即编码
#方法一
str = 'hello'
print(type(str))
byter = str.encode('utf-8')#默认为utf-8，可以换成其他编码格式，如 byter = str.encode('gb2312)
#方法二，使用codecs模块
import codecs
look = codecs.lookup('utf-8')  #建立utf-8类型的编辑器
byter = look.encode(str)
print(byter)
print(type(byter[0]))

##bytes类型转换为字符串类型，即解码
#方法一
byter = b'hello'   #建立的bytes类型，默认编码格式为ut-8
print(type(byter))
str = byter.decode('utf-8')  #如果是其他格式，则必须要改格式 
print(type(str))
#方法二
import codecs
look = codecs.lookup('utf-8')
str = look.decode(byter)
print(str)
print(type(str[0]))

<class 'str'>
(b'hello', 5)
<class 'bytes'>
<class 'bytes'>
<class 'str'>
('hello', 5)
<class 'str'>


# codecs模块讲解
* 函数一：codecs.encode(object,encoding,errors='strict')使用，errors默认为strict，还有ignore、replace等

In [96]:
import codecs
str = 'hello'
byter = codecs.encode(str,'gbk')
print(byter)
print(type(byter))

b'hello'
<class 'bytes'>


* 函数二：codecs.decode(object,encoding，errors='strict')使用,errors同encode

In [97]:
str2 = codecs.decode(byter,'gbk')
print(str2)
print(type(str2))

hello
<class 'str'>


* 函数三：codecs.lookup(encoding)使用
* 在Python编解码器注册表中查找编解码器信息，并返回如下定义的CodecInfo对象。
* encoding首先在注册表的缓存中查找。 如果未找到，则扫描已注册的搜索功能列表。 如果未找到CodecInfo对象，则引发LookupError。 否则，CodecInfo对象存储在缓存中并返回给调用者。

In [82]:
try:
    look_error = codecs.lookup('ddd') #实际并不存在ddd编码类型
except LookupError as e:
    print(e)
    
look = codecs.lookup('utf-8')#创立utf-8编码器。look是个元组，
print(look[0],look[1],look[2],look[3])
look2 = codecs.lookup('gbk')#创立gbk编码器

encoder = look[0] 
decoder = look[1]
StreamReader = look[2]
StreamWriter = look[3]

str = 'hellor'
byter = look.encode(str)
byter2 = encoder(str)
print(byter)
print(byter2)#其他用法类似

unknown encoding: ddd
<built-in function utf_8_encode> <function decode at 0x0000022F96765158> <class 'encodings.utf_8.StreamReader'> <class 'encodings.utf_8.StreamWriter'>
(b'hellor', 6)
(b'hellor', 6)


In [95]:
print(codecs.getencoder('utf-8'))
encoder = codecs.getencoder('utf-8')
byter = encoder (str)
print(byter)# 如果你只需要编码，用此法不需要创建编码器

<built-in function utf_8_encode>
(b'hello', 5)


* 以上是对codecs.getencoder(encoding)函数使用的例子，结果表示“给定编码的编码器是内置的utf-8编码函数”。其他函数的使用不再赘述。

* 函数十：codecs.register(search_function)使用：注册编解码器搜索功能。搜索函数将接受一个参数，即所有小写字母中的编码名称，并返回CodecInfo对象。如果搜索函数无法找到给定的编码，则应返回None.
* 注：搜索函数注册目前是不可逆的，这可能会在某些情况下造成问题，例如单元测试或模块重新加载。

* 函数十一：codecs.open(filename, mode='r', encoding=None, errors='strict', buffering=1)使用：encoding指定要用于文件的编码,允许对字节进行编码
和解码的任何编码，文件方法支持的数据类型取决于所使用的编解码器;errors可能会给出定义错误处理的方法,默认为'strict',如果有误，在编码时会引发
ValueError错误;buffering和内部函数open()用法一致，默认为1，即以行读取。
* 注：潜在的编码文件总是以二进制模式打开。在读取和写入的过程中，'\n'不自动转换。mode参数的用法和内置open()函数用法一致，'b'(即二进制模式)是自
动增加的

# 随笔

In [14]:
print('C:nihao\ns')#here \n means new lines
print(r'C:nihao\ns')#note the r before quote(引用)

C:nihao
s
C:nihao\ns


In [17]:
dir(a)

['__add__',
 '__class__',
 '__contains__',
 '__delattr__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__getitem__',
 '__getnewargs__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__iter__',
 '__le__',
 '__len__',
 '__lt__',
 '__mod__',
 '__mul__',
 '__ne__',
 '__new__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__rmod__',
 '__rmul__',
 '__setattr__',
 '__sizeof__',
 '__str__',
 '__subclasshook__',
 'capitalize',
 'casefold',
 'center',
 'count',
 'encode',
 'endswith',
 'expandtabs',
 'find',
 'format',
 'format_map',
 'index',
 'isalnum',
 'isalpha',
 'isdecimal',
 'isdigit',
 'isidentifier',
 'islower',
 'isnumeric',
 'isprintable',
 'isspace',
 'istitle',
 'isupper',
 'join',
 'ljust',
 'lower',
 'lstrip',
 'maketrans',
 'partition',
 'replace',
 'rfind',
 'rindex',
 'rjust',
 'rpartition',
 'rsplit',
 'rstrip',
 'split',
 'splitlines',
 'startswith',
 'strip',
 'swapcase',
 'title',
 'translate',
 'upper',
 'zfill']

In [19]:
print(type(a.encode('gb2312')))

<class 'bytes'>
